snuba/state/stateful_count.py

```diff
@@ -0,0 +1,60 @@
+from threading import Thread
+from time import time
+from typing import Any, Sequence, Tuple
+
+from arroyo.processing.strategies.dead_letter_queue import CountInvalidMessagePolicy
+from arroyo.processing.strategies.dead_letter_queue.policies.abstract import (
+    InvalidMessage,
+)
+from arroyo.types import TPayload
+
+from snuba.redis import redis_client
+
+
+class StatefulCountInvalidMessagePolicy(CountInvalidMessagePolicy[TPayload]):
+    """
+    An extension of the CountInvalidMessagePolicy which is able to save and load
+    the state of counted hits in Redis
+    """
+
+    def __init__(self, redis_hash_name: str, limit: int, seconds: int = 60) -> None:
+        self.__name = redis_hash_name
+        self.__seconds = seconds
+        super().__init__(limit, seconds, self._load_state())
+
+    def handle_invalid_message(self, e: InvalidMessage) -> None:
+        """
+        Asynchronously updates Redis hash with a new hit while
+        the base Policy handles the invalid message
+        """
+        Thread(target=self._add_to_redis).start()
+        super().handle_invalid_message(e)
+
+    def _add_to_redis(self) -> None:
+        """
+        Increments the current time entry in Redis by 1
+        """
+        now = int(time())
+        p = redis_client.pipeline()
+        p.hincrby(self.__name, str(now), 1)
+        p.hkeys(self.__name)
```

Comment:
What is this line for?

Reply:
It's to retrieve all timestamps to prune old ones. I could have just done `redis_client.hincrby()` in this method and then `redis_client.hkeys()` in the prune method but this saves 1 network call to Redis since it's done in the pipeline.